### PR TITLE
Add "Cinema" mod

### LINKED ISSUE #6029: Cinema mod is unaccessable
**Describe the bug:** You can't access Cinema mod by double clicking Auto mod.

**Screenshots or videos showing encountered issue:** https://www.youtube.com/watch?v=_CZ8KV77CHc

**osu!lazer version:** 2019.906.0

**Logs:** -

### PR DESCRIPTION
**Pull request by aspriddell**
Resolves #6029.

Add cinema mod support for all modes. In short it hides the `HUDOverlay` and `drawableRuleset.Playfield` but sets `AlwaysPresent` to `true` on each (otherwise you cannot quit the player and there will be no hitsounds)

All ruleset-specific classes have also been added to `ConvertLegacyMods` in the same way that Sudden Death and Perfect are listed (ie check for `Perfect` flag first then `else if` against the `SuddenDeath` flag)

- [x] Depends on https://github.com/ppy/osu/pull/7168

### CODE REVIEW

**Comment by GSculerlor** [comment]
Probably we can set the background dim to 0 since it kinda useless if we set dim to 100 and using Cinema mod for example (I believe stable also set the dim to 0 on Cinema mod if I'm not mistaken)

**Comment by bdach** [comment]
I'd say that's a prerequisite, given that the entire purpose of the mod is to watch the storyboard.

**Comment by aspriddell** [comment]
I'll take a look sometime tomorrow. Need to do a bit more digging...
So far I can use the configmanager to set the dim but it doesn't reset after the map has finished

**Comment by bdach** [comment]
I'm pretty sure `ConfigManager` should *not* be used here, since that might irrevocably lose user settings. Something more local would be a lot better - perhaps we'd want to add a way for mods to disable/override display settings.

**Comment by aspriddell** [comment]
I agree @bdach it seems like there is no way to revert the changes with what is currently available. 

I will look into adding an interface to deal with this (potentially called `ITempConfigOverride` or something respective)

File: osu.Game/Rulesets/Mods/ModCinema.cs [hunk 1]
```diff
@@ -3,15 +3,35 @@
...
 
 namespace osu.Game.Rulesets.Mods
 {
-    public class ModCinema : ModAutoplay
+    public abstract class ModCinema<T> : ModCinema, IApplicableToDrawableRuleset<T>
+        where T : HitObject
+    {
+        public virtual void ApplyToDrawableRuleset(DrawableRuleset<T> drawableRuleset)
+        {
+            drawableRuleset.SetReplayScore(CreateReplayScore(drawableRuleset.Beatmap));
```
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Mods/ModCinema.cs]: What about using `IApplicableFailOverride.AllowFail = false` instead?

Comment (aspriddell) [reply osu.Game/Rulesets/Mods/ModCinema.cs]: The inheritance chain should mean it already takes the settings from the Autoplay mod. `ModCinema` should inherit `IApplicableFailOverride` settings [as seen here](https://github.com/ppy/osu/blob/8da65fc9094282a458662d32c4aeb55deccfa721/osu.Game/Rulesets/Mods/ModAutoplay.cs#L21)

File: osu.Game/Screens/Play/Player.cs [hunk 2]
```diff
@@ -498,8 +498,24 @@ public override void OnEntering(IScreen last)
                 .Delay(250)
                 .FadeIn(250);
 
-            Background.EnableUserDim.Value = true;
-            Background.BlurAmount.Value = 0;
+            var screenOverride = Mods.Value.OfType<IApplicableToScreen>();
```
Comment (aspriddell) [inline osu.Game/Screens/Play/Player.cs]: Was thinking that if more than one mod in the list had the `IApplicableToScreen` modifier it threw an exception. In this case line 503 would become a switch with `0`, `1` and `default`, with default throwing the exception.

Any ideas on an appropriate exception to throw, or should I make my own, like `IncompatableModSelection`?

File: osu.Game/Rulesets/Mods/IApplicableToScreen.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// An interface for a mod which can temporarily override screen settings.
+    /// </summary>
+    public interface IApplicableToScreen : IApplicableMod
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/IApplicableToScreen.cs]: this interface is too specific. you should be providing access to the required instances, not exposing the three things specifically required by cinema mod alone.

File: osu.Game/Screens/Play/Player.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: there's already a system in place for incompatible mods (see `IncompatibleMods`). this kind of logic should not exist.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Mods/IApplicableToPlayer.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// An interface for a mod which can temporarily override the <see cref="Player"/> settings.
+    /// </summary>
+    public interface IApplicableToPlayer : IApplicableMod
+    {
+        void ApplyToPlayer(Player player);
```
Comment (aspriddell) [inline osu.Game/Rulesets/Mods/IApplicableToPlayer.cs]: Much less specific now, but can still do everything from before

**Comment by peppy** [comment]
If we are using the `IgnoreUserSettings` bindable, it needs test coverage in `TestSceneUserDimContainer`.

File: osu.Game/Graphics/Containers/UserDimContainer.cs [hunk 5]
```diff
@@ -90,7 +91,7 @@ protected virtual void UpdateVisuals()
             ContentDisplayed = ShowDimContent;
 
             dimContent.FadeTo(ContentDisplayed ? 1 : 0, BACKGROUND_FADE_DURATION, Easing.OutQuint);
-            dimContent.FadeColour(OsuColour.Gray(1 - (float)DimLevel), BACKGROUND_FADE_DURATION, Easing.OutQuint);
+            dimContent.FadeColour(IgnoreUserSettings.Value ? OsuColour.Gray(1 - (float)DimLevel) : Color4.White, BACKGROUND_FADE_DURATION, Easing.OutQuint);
```
Comment (peppy) [inline osu.Game/Graphics/Containers/UserDimContainer.cs]: Is this line required? Shouldn't it be handled by the `DimLevel` one above?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Graphics/Containers/UserDimContainer.cs (hunk 5 above)
Comment (aspriddell) [reply osu.Game/Graphics/Containers/UserDimContainer.cs]: resolved in caa9286

File: osu.Game/Rulesets/Mods/ModCinema.cs [hunk 6]
```diff
@@ -3,15 +3,43 @@
...
+
+            drawableRuleset.Playfield.AlwaysPresent = true;
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/ModCinema.cs]: Can you explain the reason behind this being required? I'm going through and trying to remove this requirement at the `Player` level.

Comment (peppy) [reply osu.Game/Rulesets/Mods/ModCinema.cs]: Required for hitsounds.

**Comment by peppy** [comment]
I've made a few changes:

- Also hides break overlay
- Removes AlwaysPresent requirement on HUD overlay via https://github.com/ppy/osu/pull/7168
- Adds a comment about playfield `AlwaysPresent` requirement

**Comment by aspriddell** [comment]
👍

**Comment by peppy** [comment]
This is the absolute last change to be made to `UserDimContainer`. That class needs a rewrite to remove the 10+ settings it has. I'm OK with this happening so we at least have documentation of all the different parameters we need to account for when refactoring the class.

**Review by peppy — APPROVED** [review]

**Pull request merged**